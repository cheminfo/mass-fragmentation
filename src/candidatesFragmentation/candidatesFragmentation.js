import MassTools from 'mass-tools';
import generateMFs from 'mf-generator';
import OCL from 'openchemlib';
import { getMF } from 'openchemlib-utils';

import { bondContribution } from '../bondContribution/bondContribution.js';
import { fragment } from '../fragmentation/fragment.js';
import { neutralLoss } from '../neutralLoss/neutralLoss.js';

const { Spectrum } = MassTools;
const { Molecule } = OCL;

/**
 * This function performs the in-silico fragmentation of a given structure and returns the matched fragments and the bond contribution
 * @param {object} spectrum experimental spectra in form {x:[],y:[]}
 * @param {any} idCode The OCL idCode of candidate structure
 * @param {object} options Object containing the ionization (like H+, +,..) and the precision in ppm (like 5ppm) in a format {ionization:[H+],precision:[5ppm]}
 * @returns returns matched fragments with their bond contribution on experimental spectra
 */

export async function candidatesFragmentation(spectrum, idCode, options) {
  const { precision, ionization, limit } = options;

  // Generate molecule from smiles
  const molecule = Molecule.fromIDCode(idCode);

  // Get mass precursor ion, used in bondContribution() and MF molecular
  let mfMolecularIon = getMF(molecule).mf;
  let mfPrecursorIon = await generateMFs([mfMolecularIon], {
    ionizations: ionization,
  });
  let massPrecursorIon = mfPrecursorIon[0].ms.em;

  // Filtration of experimental spectrum with Mf of molecular ion
  let newSpectrum = new Spectrum(spectrum);
  let filteredSpectrum = await newSpectrum.getFragmentPeaks(mfMolecularIon, {
    ionizations: ionization,
    precision,
  });
  // filteredSpectrumMasses used for matching in generateMFs() and filtredSpectrumFroStatistics in format for bondContribution() and for matching
  const masses = filteredSpectrum.map((peak) => peak.x);

  // Perform in-silico fragmentation
  const fragmentation = fragment(molecule);

  let mfsArray = [];
  const optionsMFs = {
    ionizations: ionization,
    limit,
    uniqueMFs: false,
    filter: {
      targetMasses: masses,
      precision,
    },
  };
  let fragmentsResult = [];
  for (let i = 0; i < fragmentation.length; i++) {
    if (fragmentation[i].hose !== undefined) {
      // get neutral losses for each fragment
      let neutralLosses = neutralLoss(fragmentation[i].idCode);
      mfsArray[i] = [fragmentation[i].mf, neutralLosses];
      let results = await generateMFs(mfsArray[i], optionsMFs);
      let groups = {};
      for (const result of results) {
        const em = Math.round(result.ms.em * 1e6);
        if (!groups[em]) {
          groups[em] = {
            em: result.em,
            ms: result.ms.em,
            mf: result.mf,
            ppm: result.ms.ppm,
            mfs: [],
            hose: fragmentation[i].hose,
            ignored: 0,
          };
        } else {
          groups[em].ignored++;
        }
        groups[em].mfs.push(result.mf);
      }

      groups = Object.values(groups);

      if (groups.length > 0) {
        for (let j = 0; j < groups.length; j++) {
          fragmentsResult.push(groups[j]);
        }
      }
    }
    // Account for Molecular ion to be matched
    if (fragmentation[i].fragmentType === 'Molecular Ion') {
      let resultMolecularIon = await generateMFs(
        [fragmentation[i].mf],
        optionsMFs,
      );
      let group = {};
      for (const result of resultMolecularIon) {
        const em = Math.round(result.ms.em * 1e6);

        group[em] = {
          em: result.em,
          ms: result.ms.em,
          mf: result.mf,
          ppm: result.ms.ppm,
          mfs: [],
          hose: fragmentation[i].hose,
          ignored: 0,
        };

        group[em].mfs.push(result);
      }

      group = Object.values(group);

      if (group.length > 0) {
        for (let j = 0; j < group.length; j++) {
          fragmentsResult.push(group[j]);
        }
      }
    }
  }

  let x = [];
  let y = [];
  for (let i = 0; i < filteredSpectrum.length; i++) {
    x.push(filteredSpectrum[i].x);
    y.push(filteredSpectrum[i].y);
  }
  let spectrumForContribution = { x, y };
  // get intensity of each matched fragment
  for (let i = 0; i < fragmentsResult.length; i++) {
    let massAccuracy = (precision * fragmentsResult[i].ms) / 1e6;
    for (let m = 0; m < filteredSpectrum.length; m++) {
      if (
        Math.abs(filteredSpectrum[m].x - fragmentsResult[i].ms) <= massAccuracy
      ) {
        fragmentsResult[i].intensity = filteredSpectrum[m].y;
      }
      if (fragmentsResult[i].hose === undefined) {
        fragmentsResult[i].intensity = filteredSpectrum[m].y;
      }
    }
  }
  // Bond Contribution of each matched fragment

  let resultContribution = bondContribution(
    spectrumForContribution,
    massPrecursorIon,
    precision,
  );
  //console.log(resultContribution);

  if (resultContribution.length > 0) {
    for (let i = 0; i < fragmentsResult.length; i++) {
      let massAccuracyOfFragment =
        (options.precision * fragmentsResult[i].ms) / 1e6;
      for (let j = 0; j < resultContribution.length; j++) {
        if (
          Math.abs(resultContribution[j].mass - fragmentsResult[i].ms) <=
          massAccuracyOfFragment
        ) {
          fragmentsResult[i].contribution = resultContribution[j].contribution;
        }
        if (fragmentsResult[i].hose === undefined) {
          fragmentsResult[i].contribution = 0;
        }
      }
    }
  }
  return fragmentsResult;
}
