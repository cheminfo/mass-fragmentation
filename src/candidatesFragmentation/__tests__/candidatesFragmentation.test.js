import { toBeDeepCloseTo, toMatchCloseTo } from 'jest-matcher-deep-close-to';

import { candidatesFragmentation } from '../candidatesFragmentation.js';

expect.extend({ toBeDeepCloseTo, toMatchCloseTo });

describe('fragment', () => {
  it('CC(=O)CCC(=O)O', async () => {
    const spectrum = {
      x: [
        15.02348, 42.01056, 46.00548, 59.0133, 57.03404, 71.04968, 101.02386,
        117.05462,
      ],
      y: [50, 80, 40, 20, 60, 20, 30, 100],
    };
    const smiles = 'CC(=O)CCC(=O)O';
    const options = { precision: 15, ionization: 'H+', limit: 1e7 };
    const result = await candidatesFragmentation(spectrum, smiles, options);
    expect(result[0]).toMatchCloseTo({
      contribution: 0,
      em: 116.04734411655001,
      hose: undefined,
      intensity: 100,
      mf: 'C5H8O3',
      mfs: [
        {
          atoms: { C: 5, H: 8, O: 3 },
          charge: 0,
          em: 116.04734411655001,
          ionization: {
            atoms: { H: 1 },
            charge: 1,
            em: 1.00782503223,
            mf: 'H+',
          },
          mf: 'C5H8O3',
          ms: {
            charge: 1,
            delta: -5.688709450168972e-7,
            em: 117.05462056887094,
            ionization: 'H+',
            ppm: -0.004859876032678206,
            target: { mass: 117.05462 },
          },
          mw: 116.11542028907729,
          parts: ['C5H8O3'],
          unsaturation: 2,
        },
      ],
      ms: 117.05462056887094,
      ppm: -0.004859876032678206,
    });
  });
  it('CC(N)C1CCC(CC#N)C(CC=O)C1', async () => {
    const spectrum = {
      x: [
        45.05785, 71.04969, 143.13101, 165.11536, 193.14666, 194.14191,
        209.16539,
      ],
      y: [50, 80, 40, 20, 60, 20, 100],
    };
    const smiles = 'CC(N)C1CCC(CC#N)C(CC=O)C1';
    const options = { precision: 15, ionization: 'H+', limit: 1e7 };
    const result = await candidatesFragmentation(spectrum, smiles, options);
    expect(result[0]).toMatchCloseTo({
      contribution: 0,
      em: 208.15756327303,
      hose: undefined,
      intensity: 100,
      mf: 'C12H20N2O',
      mfs: [
        {
          atoms: { C: 12, H: 20, N: 2, O: 1 },
          charge: 0,
          em: 208.15756327303,
          ionization: {
            atoms: { H: 1 },
            charge: 1,
            em: 1.00782503223,
            mf: 'H+',
          },
          mf: 'C12H20N2O',
          ms: {
            charge: 1,
            delta: 0.0005502746490719801,
            em: 209.16483972535093,
            ionization: 'H+',
            ppm: 2.6308181135726816,
            target: { mass: 209.16539 },
          },
          mw: 208.30045718914837,
          parts: ['C12H20N2O'],
          unsaturation: 4,
        },
      ],
      ms: 209.16483972535093,
      ppm: 2.6308181135726816,
    });
  });
});
