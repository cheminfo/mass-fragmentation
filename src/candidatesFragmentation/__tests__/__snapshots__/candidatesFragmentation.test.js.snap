// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment CC(=O)CCC(=O)O 1`] = `
[
  {
    "contribution": 0,
    "em": 116.04734411655001,
    "hose": undefined,
    "intensity": 100,
    "mf": "C5H8O3",
    "mfs": [
      {
        "atoms": {
          "C": 5,
          "H": 8,
          "O": 3,
        },
        "charge": 0,
        "em": 116.04734411655001,
        "ionization": {
          "atoms": {
            "H": 1,
          },
          "charge": 1,
          "em": 1.00782503223,
          "mf": "H+",
        },
        "mf": "C5H8O3",
        "ms": {
          "charge": 1,
          "delta": -5.688709450168972e-7,
          "em": 117.05462056887094,
          "ionization": "H+",
          "ppm": -0.004859876032678206,
          "target": {
            "mass": 117.05462,
          },
        },
        "mw": 116.11542028907729,
        "parts": [
          "C5H8O3",
        ],
        "unsaturation": 2,
      },
    ],
    "ms": 117.05462056887094,
    "ppm": -0.004859876032678206,
  },
  {
    "contribution": 0.16666666666666666,
    "em": 100.01604398763,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gCa@@dsPGBe@jCtW_hdcaUIpj\`",
        "gJQ@@dvu@\\JTBhOI]lQIGBjSaU@",
        "gGQ@@dvmTApiPJ\`}evpy~bRNETgBj@",
        "daxL@@[dfufjX@xRJAJCvIvXV_hbPxURNET",
      ],
    },
    "intensity": 30,
    "mf": "C4H4O3",
    "mfs": [
      "C4H4O3",
    ],
    "ms": 101.02332043995094,
    "ppm": 5.340945503604677,
  },
  {
    "contribution": 1,
    "em": 41.002739651800006,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJQ@@dru@\\JTBhOQY~bRNETgBj@",
        "gGQ@@dsmTApiPJ\`}efxy~bRNETgBj@",
        "daxL@@[df]fjX@xRJAJCvIf\\V_hbPxURNET",
        "daxL@@[df]fjX@XRJAJC}DRGBjQpj\`",
      ],
    },
    "intensity": 80,
    "mf": "C2HO",
    "mfs": [
      "C2HO",
    ],
    "ms": 42.010016104120936,
    "ppm": 12.946814343369255,
  },
  {
    "contribution": 0.6666666666666666,
    "em": 56.02621474848999,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
        "daxL@@QdfufjX@xSJ@jCrYaY~bIcaUDxUP",
        "daxL@@QdfufjX@XSJ@jC}DSGBjIpj\`",
        "daxL@@QdfufjX@XSJ@jC}DSGBjIpj\`",
      ],
    },
    "intensity": 60,
    "mf": "C3H4O",
    "mfs": [
      "C3H4O",
    ],
    "ms": 57.03349120081092,
    "ppm": 9.622402162682588,
  },
  {
    "contribution": 0,
    "em": 58.0054793036,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
        "daxL@@QdfufjX@xSJ@jCrYaY~bIcaUDxUP",
        "daxL@@QdfufjX@XSJ@jC}DSGBjIpj\`",
        "daxL@@QdfufjX@XSJ@jC}DSGBjIpj\`",
      ],
    },
    "intensity": 20,
    "mf": "C2H2O2",
    "mfs": [
      "C2H2O2",
    ],
    "ms": 59.01275575592093,
    "ppm": 9.222482022769242,
  },
  {
    "contribution": 0,
    "em": 70.04186481295,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJP\`@dfvhCaR\`uAz[OtRQpjlxUP",
        "gGP\`@dfuj\`NEJCTGlhvGOtRQpjlxUP",
        "daxL@@QdfufjX@xRJCJCvIFXV_hbPxUVNET",
        "daxL@@QdfufjX@XRJCJC}DRGBjqpj\`",
      ],
    },
    "intensity": 20,
    "mf": "C4H6O",
    "mfs": [
      "C4H6O",
    ],
    "ms": 71.04914126527093,
    "ppm": 7.5825649609296555,
  },
  {
    "contribution": 0.33333333333333337,
    "em": 44.99765427137,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJP\`@dfvhCaR\`uAz[OtRQpjlxUP",
        "gGP\`@dfuj\`NEJCTGlhvGOtRQpjlxUP",
        "daxL@@QdfufjX@xRJCJCvIFXV_hbPxUVNET",
        "daxL@@QdfufjX@XRJCJC}DRGBjqpj\`",
      ],
    },
    "intensity": 40,
    "mf": "CHO2",
    "mfs": [
      "CHO2",
    ],
    "ms": 46.00493072369093,
    "ppm": 11.93950953572249,
  },
]
`;

exports[`fragment CC(N)C1CCC(CC#N)C(CC=O)C1 1`] = `
[
  {
    "contribution": 0,
    "em": 208.15756327303,
    "hose": undefined,
    "intensity": 100,
    "mf": "C12H20N2O",
    "mfs": [
      {
        "atoms": {
          "C": 12,
          "H": 20,
          "N": 2,
          "O": 1,
        },
        "charge": 0,
        "em": 208.15756327303,
        "ionization": {
          "atoms": {
            "H": 1,
          },
          "charge": 1,
          "em": 1.00782503223,
          "mf": "H+",
        },
        "mf": "C12H20N2O",
        "ms": {
          "charge": 1,
          "delta": 0.0005502746490719801,
          "em": 209.16483972535093,
          "ionization": "H+",
          "ppm": 2.6308181135726816,
          "target": {
            "mass": 209.16539,
          },
        },
        "mw": 208.30045718914837,
        "parts": [
          "C12H20N2O",
        ],
        "unsaturation": 4,
      },
    ],
    "ms": 209.16483972535093,
    "ppm": 2.6308181135726816,
  },
  {
    "contribution": 0,
    "em": 164.10753907522002,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gCh@@duTe@xThEP^Ta]~bRNETgBj@",
        "gGX@@dwUUIPNEJATGbLdKmGzIHxUR\\Jh",
        "daz@@@R[ejjhb\`NDb\`R\`|pYDIvPVUgzHdNETcaU@",
        "deV@@@R[e[ajjjlbcJFHCaHhDhOBFQB]dEeXt]gzHdNETcaU@",
      ],
    },
    "intensity": 20,
    "mf": "C10H14NO",
    "mfs": [
      "C10H14NO",
    ],
    "ms": 165.11481552754094,
    "ppm": 3.2975384875625484,
  },
  {
    "contribution": 0,
    "em": 193.13408817634001,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gCh@@duTe@xThEP^Ta]~bRNETgBj@",
        "gGX@@dwUUIPNEJATGbLdKmGzIHxUR\\Jh",
        "daz@@@R[ejjhb\`NDb\`R\`|pYDIvPVUgzHdNETcaU@",
        "deV@@@R[e[ajjjlbcJFHCaHhDhOBFQB]dEeXt]gzHdNETcaU@",
      ],
    },
    "intensity": 20,
    "mf": "C11H17N2O",
    "mfs": [
      "C11H17N2O",
    ],
    "ms": 194.14136462866094,
    "ppm": 2.8091454909730884,
  },
  {
    "contribution": 0.6666666666666666,
    "em": 192.13883920414003,
    "hose": {
      "bond1": [
        "eF\`BLGBTEP_hhKeUNET",
        "gCh@@duPGBEAJCrFROtRA\\jhxUP",
        "gGX@@dsUUIPNDJBTGbLdJMwzI@nUT\\Jh",
        "daz@@@RYmjjhb\`NDB\`b\`|pYDIF\\VUgzH\`KeUCaU@",
        "deV@@@RYm[ajjjlbaBFhCa@hHhOBFQBQgEeXt]gzH\`KeUCaU@",
      ],
    },
    "intensity": 60,
    "mf": "C12H18NO",
    "mfs": [
      "C12H18NO",
    ],
    "ms": 193.14611565646095,
    "ppm": 2.818299178305648,
  },
  {
    "contribution": 0.5000000000000001,
    "em": 44.05002419781,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gGX@@dsUUIPNEJATGeHT_hdcaUIpj\`",
        "daz@@@RYmjjhb\`NDb\`R\`|pYDIF\\VUgzHdNETcaU@",
        "deV@@@RYm[ajjjl\`aJFHCaHhDhOBFQBQgEeXt]gzHdNETcaU@",
        "dcN@@@RYm[^FjjjhhHRabAHCaHhDhOJFQBQgEeXt]\`qN_hbPxURNET",
      ],
    },
    "intensity": 50,
    "mf": "C2H6N",
    "mfs": [
      "C2H6N",
    ],
    "ms": 45.05730065013093,
    "ppm": 12.192249893998104,
  },
  {
    "contribution": 0,
    "em": 164.10753907522002,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gGX@@dsUUIPNEJATGeHT_hdcaUIpj\`",
        "daz@@@RYmjjhb\`NDb\`R\`|pYDIF\\VUgzHdNETcaU@",
        "deV@@@RYm[ajjjl\`aJFHCaHhDhOBFQBQgEeXt]gzHdNETcaU@",
        "dcN@@@RYm[^FjjjhhHRabAHCaHhDhOJFQBQgEeXt]\`qN_hbPxURNET",
      ],
    },
    "intensity": 20,
    "mf": "C10H14NO",
    "mfs": [
      "C10H14NO",
    ],
    "ms": 165.11481552754094,
    "ppm": 3.2975384875625484,
  },
  {
    "contribution": 0,
    "em": 164.10753907522002,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJP@DjZhCaB\`eAyBIgzI@xUT\\Jh",
        "dif@@@RYm^FjjiHHP\`NDR\`J\`}TPdEeXv_hbHxUQNET",
        "dmv@@@RYmUxZjjjJBDHx\`Z@xQJ@jCtadPdYqYVMgPL_hbHxUQNET",
        "dkn@@@RYmUWajjjjb\`aBNHF\`NDR\`J\`}hYDIF\\VUcYtCDxn_hbHxUQNET",
      ],
      "bond2": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJP@DjZhCaB\`eAyBIgzI@xUT\\Jh",
        "did@@DjVVajjjPJDhCa@hHhOTDQbQfEgzH\`NEUCaU@",
        "dcN@@@RUeg^gjjjkLhjbR@xSJ@jCpifPVUCYvCdq~bIcaUDxUP",
        "do}H@ClDeYeWXYZjjkjKJBHFcZ@xPjAjCrqBYdEEXv]\`qN[GzHaNETsaU@",
      ],
    },
    "intensity": 20,
    "mf": "C10H14NO",
    "mfs": [
      "C10H14NO",
    ],
    "ms": 165.11481552754094,
    "ppm": 3.2975384875625484,
  },
  {
    "contribution": 0.33333333333333337,
    "em": 142.12318913968002,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
        "daD@@DiYbZjjdQY@xPJBJCqAdXdYGzH\`NEUCaU@",
        "dcN@@@RYmU~FjjjhhHPcbAhCaBhFhOCDIAYVMgPLKgzHaNETsaU@",
        "dg]@@@MIVV]{izjjjhlhjbRMhCaBhFhOZfYAYTMgXNSFp^WwzHaNETsaU@",
      ],
      "bond2": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gGP@DjZj\`NDJBTGdHd_hdCaUQpj\`",
        "deT@@DjYUXZjji@TPNDB\`b\`|pQDIfXVUgzH\`NEUCaU@",
        "dknH@CAIVYUqejjjnsBBhD\`NDr\`J\`}LPfYAQVMgXLSgzHfNETSaU@",
        "do}H@ClDeYeWXYZjjkjKBBhDcR@xSJ@jCrqBYdEEXv]\`qN[GzHfNETSaU@",
      ],
    },
    "intensity": 40,
    "mf": "C8H16NO",
    "mfs": [
      "C8H16NO",
    ],
    "ms": 143.13046559200095,
    "ppm": 3.803578761561851,
  },
  {
    "contribution": 1,
    "em": 70.04186481295,
    "hose": {
      "bond1": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gGP@DjZj\`NDJBTGdHd_hdCaUQpj\`",
        "deT@@DjYUXZjji@TPNDB\`b\`|pQDIfXVUgzH\`NEUCaU@",
        "dknH@CAIVYUqejjjnsBBhD\`NDr\`J\`}LPfYAQVMgXLSgzHfNETSaU@",
        "do}H@ClDeYeWXYZjjkjKBBhDcR@xSJ@jCrqBYdEEXv]\`qN[GzHfNETSaU@",
      ],
      "bond2": [
        "eF@Hp\\IPUA~b\`xUTxUP",
        "gJP@DjZhCaB\`eAyBIgzI@xUT\\Jh",
        "deV@@@RYm[ajjjlbaBFhCaDhBhOUDIAYVMGzHbNETSaU@",
        "dcN@@@RYm[^FjjjhhhPajAhCaDhBhOJFQBQgEeXt]\`qN_hbHxUQNET",
        "dg^H@CAIVV[[eZjjjhlhjajIHCaBhFhOGDIfPVUCQvCdpnGe}~bHSaULxUP",
      ],
    },
    "intensity": 80,
    "mf": "C4H6O",
    "mfs": [
      "C4H6O",
    ],
    "ms": 71.04914126527093,
    "ppm": 7.723312615652714,
  },
]
`;
